sw/device/mask_rom/boot.md

```diff
@@ -1,74 +1,182 @@
+
+   **Open Q:** Whether SW has to configure PMP initial region.
 
 2. Execution Begins
-  - CRT Startup Code (ASM)
-    - Clean Slate (SRAM, Registers)
-    - Well-defined interrupt/exception handler
-        Open Q: Do we want to enable interrupts in mask rom?
-    - Jump into C code (`boot`):
+   *   CRT Startup Code (Written in Assembly)
+
+       *   Disable Interrupts and set well-defined exception handler.
+           This should keep initial execution deterministic.
+
+       *   Clean Device State Part 1. (See "Cleaning Device State" below)
+
+       *   Setup structures for C execution (CRT: `.data`, `.bss` sections, stack).
+
+       *   Jump into C code
+
+   *   Main Secure Boot Software (Written in C)
+       *   Orchestrated from `boot`:
 
 ```c
 void boot(void) {
-  // - Clean all other hardware state (using DIF Reset functions)
-  clean_device_state();
-
-  // - Chip-specific startup functionality
-  //   - Clocks
-  //   - AST
-  //   - Flash
-  //   - Entropy
-  //   Open Q: Proprietary Software Strategy.
+  // Clean Device State Part 2.
+  // See "Cleaning Device State" Below.
+  clean_device_state_part_2();
+
+  // Enable Memory Protection
+  // - SRAM Scramble (may have to happen earlier)
+  // - PMP Initial Region (if not done in power on)
+  enable_memory_protection();
+
+  // Chip-specific startup functionality
+  // **Open Q:** Proprietary Software Strategy.
+  // - Clocks
+  // - AST
+  // - Flash
+  // - Fuses
+  // - Entropy
   chip_specific_startup();
 
-  // - Read Boot Policy for ROM_EXT from Flash (2.b)
-  //   Open Q: How is this protected beyond being in flash?
-  boot_policy = read_boot_policy();
+  // Manufacturing boot-strapping intervention.
+  // **Open Q:** How does the mask rom deal with chip recovery?
+  manufacturing_boot_strap();
 
-  // - Determine which rom_ext slot is in use (2.b,2.c.i)
-  rom_ext_manifest = read_rom_ext_slot(boot_policy);
+  // Read Boot Policy for ROM_EXT from Flash (2.b)
+  // **Open Q:** How is this protected beyond being in flash?
+  boot_policy = read_boot_policy();
 
-  // Verify ROM_EXT Image Digest (2.c.ii)
-  if (!verify_rom_ext_hash(rom_ext_manifest)) {
-    // Boot Failure (check policy)
+  // Determine which ROM_EXT slot is prioritised (2.b, 2.c.i)
+  for ( current_rom_ext_manifest in rom_ext_manifests_to_try(boot_policy) ) {
+
+    // Sanity Check ROM_EXT Manifest (2.c.ii)
+    // **Open Q:** Integration with Secure Boot Hardware
+    // - Header Format
+    // - Plausible Key
+    // - Initial Digest Checks
+    if (!check_rom_ext_manifest(current_rom_ext_manifest)) {
+      // Boot Failure (check policy)
+      if (try_next_on_manifest_failed(boot_policy))
+        continue
+      else
+        break
+    }
+
+    // Verify ROM_EXT Image Signature (2.c.iii)
+    // **Open Q:** Integration with Secure Boot Hardware, OTBN
+    // **Open Q:** Key Selection method/mechanism.
+    valid, signature = verify_rom_ext_signature_start(pub_key_selector(...),
```

Comment:
nit: s/signature/hash_padded_result/ or something similar? or remove if just planning to use the value provided by `verify_rom_ext_signature_result`

Reply:
I've dropped it in favour of `verify_rom_ext_signature_result`, but also consider that the Open Question about Secure Boot Hardware to answer how we need to handle state coming from this function.